### PR TITLE
[297] Record trainee ethnic background

### PR DESCRIPTION
**Pull request by darokel**
### Context

- https://trello.com/c/8hPHg3yz/297-m-record-trainee-ethnic-background

### Changes proposed in this pull request

- Implement the ethnic background section in the diversity information flow

Screenshot 2020-10-28 at 11 42 02

Screenshot 2020-10-28 at 11 42 08

### Guidance to review

- Visit the review app, diversity section - https://dfe-twd-rttd-pr-119.herokuapp.com/
- Begin the flow all the way up to ethnic background
- Select a background and assert it is saved properly
- Return and choose a different ethnic group, the previous background should be cleared allowing you to choose a new one.

### CODE REVIEW

**Comment by defong** [comment]
> Kudos, SonarCloud Quality Gate passed!
> 
> [Bug](https://sonarcloud.io/project/issues?id=DFE-Digital_register-trainee-teacher-data&pullRequest=119&resolved=false&types=BUG) [A](https://sonarcloud.io/project/issues?id=DFE-Digital_register-trainee-teacher-data&pullRequest=119&resolved=false&types=BUG) [0 Bugs](https://sonarcloud.io/project/issues?id=DFE-Digital_register-trainee-teacher-data&pullRequest=119&resolved=false&types=BUG)
> [Vulnerability](https://sonarcloud.io/project/issues?id=DFE-Digital_register-trainee-teacher-data&pullRequest=119&resolved=false&types=VULNERABILITY) [A](https://sonarcloud.io/project/issues?id=DFE-Digital_register-trainee-teacher-data&pullRequest=119&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=DFE-Digital_register-trainee-teacher-data&pullRequest=119&resolved=false&types=VULNERABILITY) (and [Security Hotspot](https://sonarcloud.io/project/issues?id=DFE-Digital_register-trainee-teacher-data&pullRequest=119&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/issues?id=DFE-Digital_register-trainee-teacher-data&pullRequest=119&resolved=false&types=SECURITY_HOTSPOT) to review)
> [Code Smell](https://sonarcloud.io/project/issues?id=DFE-Digital_register-trainee-teacher-data&pullRequest=119&resolved=false&types=CODE_SMELL) [A](https://sonarcloud.io/project/issues?id=DFE-Digital_register-trainee-teacher-data&pullRequest=119&resolved=false&types=CODE_SMELL) [1 Code Smell](https://sonarcloud.io/project/issues?id=DFE-Digital_register-trainee-teacher-data&pullRequest=119&resolved=false&types=CODE_SMELL)
> 
> [No Coverage information](https://sonarcloud.io/component_measures?id=DFE-Digital_register-trainee-teacher-data&pullRequest=119) No Coverage information
> [](https://sonarcloud.io/component_measures?id=DFE-Digital_register-trainee-teacher-data&pullRequest=119&metric=new_duplicated_lines_density&view=list) [0.0% Duplication](https://sonarcloud.io/component_measures?id=DFE-Digital_register-trainee-teacher-data&pullRequest=119&metric=new_duplicated_lines_density&view=list)

image

**Comment by darokel** [comment]
Apologies for the slightly chunky PR but in doing this section, I had to refactor and clean up the code for the previous sections to keep things consistent. Wanted to leave this bit till the last diversity section had been finished but this required me to do a bit of house keeping now.

**Comment by defong** [comment]
[[M] Record trainee ethnic background](https://trello.com/c/8hPHg3yz/297-m-record-trainee-ethnic-background)

File: app/controllers/trainees/diversity/disclosures_controller.rb [hunk 1]
```diff
@@ -7,9 +7,10 @@ def edit
 
       def update
         updater = Diversities::Disclosures::Update.call(trainee: trainee, attributes: disclosure_param)
+        trainee = updater.disclosure.trainee
```
Comment (defong) [inline app/controllers/trainees/diversity/disclosures_controller.rb:10]: can we move that closer to usage, there is a private `trainee` method...

File: app/controllers/trainees/diversity/ethnic_groups_controller.rb [hunk 2]
```diff
@@ -7,9 +7,10 @@ def edit
 
       def update
         updater = Diversities::EthnicGroups::Update.call(trainee: trainee, attributes: ethnic_group_param)
+        trainee = updater.ethnic_group.trainee
```
Comment (defong) [inline app/controllers/trainees/diversity/ethnic_groups_controller.rb:10]: can we move that closer to usage, there is a private `trainee` method...

File: spec/features/trainees/diversities/edit_ethnic_group_spec.rb [hunk 3]
```diff
@@ -17,36 +17,87 @@
...
   def then_i_see_error_messages
     expect(page).to have_content(
       I18n.t("activemodel.errors.models.diversities/ethnic_group.attributes.ethnic_group.blank"),
     )
   end
+
+private
+
+  def ethnic_background_option(ethnic_background)
+    formatted_id = ethnic_background.downcase.gsub(" ", "-")
```
Comment (defong) [inline spec/features/trainees/diversities/edit_ethnic_group_spec.rb:100]: maybe 
```suggestion
    formatted_id = ethnic_background.parameterize
```

https://apidock.com/rails/String/parameterize

File: spec/services/diversities/ethnic_backgrounds/update_spec.rb [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        before(:each) do
+          service.call
+          trainee.reload
+        end
+
+        context "when ethnic background attributes are valid" do
+          let(:attributes) do
+            {
+              ethnic_background: "Another ethnic background",
+              additional_ethnic_background: "some other bacground",
```
Comment (defong) [inline spec/services/diversities/ethnic_backgrounds/update_spec.rb:19]:
```suggestion
              additional_ethnic_background: "some other background",
```

**Review by defong — APPROVED** [review]

File: app/controllers/trainees/diversity/disclosures_controller.rb (hunk 1 above)
Comment (darokel) [reply app/controllers/trainees/diversity/disclosures_controller.rb]: done https://github.com/DFE-Digital/register-trainee-teachers/pull/119/commits/8e7814211ba680f2b9d0e1f6b510a84c046ba925, cleaned up a few other bits too

File: app/controllers/trainees/diversity/ethnic_groups_controller.rb (hunk 2 above)
Comment (darokel) [reply app/controllers/trainees/diversity/ethnic_groups_controller.rb]: done https://github.com/DFE-Digital/register-trainee-teachers/pull/119/commits/8e7814211ba680f2b9d0e1f6b510a84c046ba925, cleaned up a few other bits too

File: spec/features/trainees/diversities/edit_ethnic_group_spec.rb [hunk 5]
```diff
@@ -45,8 +54,10 @@ def when_i_visit_the_diversity_ethnic_group_page
     @ethnic_group_page.load(id: @trainee.id)
   end
 
-  def and_i_enter_valid_parameters
-    @ethnic_group_page.asian.choose
+  def and_i_choose(option)
+    @ethnic_group_page.find(
+      "#diversities-ethnic-group-ethnic-group-#{option.dasherize}-field",
```
Comment (aliuk2012) [inline spec/features/trainees/diversities/edit_ethnic_group_spec.rb:59]: this could be a site prism object helper method

**Review by aliuk2012 — APPROVED** [review]

**Pull request merged**